test/unit/org/apache/cassandra/index/sai/functional/CompactionTest.java

```diff
@@ -155,6 +156,20 @@ public void testConcurrentQueryWithCompaction() throws Throwable
         verifySSTableIndexes(v2IndexName, num);
     }
 
+    @Test
+    public void testCompactionWithDisabledIndexReads() throws Throwable
```

Comment:
This test is sort of confusing: it passes, but one would expect to fail given reads are disabled? Or to rephrase, can we make a test that specifically checks for that?

Reply:
You're right there was something fishy here. I had misread the test I was reusing, but it shouldn't have passed, and the reason it did is that the original test was kind of broken. More precisely, the `TestWithConcurrentVerification` utility used by a few of the tests of this file was not really working correctly: if the verification was throwing, the exception was logged but otherwise swallowed and didn't fail the test. In fact, there was a syntax error in the queries used by the existing `testConcurrentQueryWithCompaction`, so the verification task *was* failing, but the test wasn't.

Now, I fixed that utility in the last commit (https://github.com/datastax/cassandra/commit/07439b46189b4504f7127c24c7058e5f35501e8e) but this highlighted a few things:
1. the idea of throwing an exception if search is used doesn't quite work: the `SearchableIndex#search` method isn't called because the rest of the methods (`minTerm`, `maxTerm`, ...) already tell the planner that there is nothing to do. Now, just having all methods of the `SearchableIndex` also doesn`t work easily because a number of those min/max methods are used when opening the sstable, not just when "reading" the index. There may also be way to trick the planner to force calling the `search` method, but I don't know the planner at all and I don't feel spending 2 days on understanding it enough is a good use of time here. So I modified the patch to simply document that "disabling index reads" simply means queries will always return an empty result, which should be find in practice.
2. unfortunately, fixing `TestWithConcurrentVerification` meant that one of the existing test, `testConcurrentIndexBuildWithCompaction`, started failing. I looked a bit, and tbh I couldn't quite figure what the test was trying to do in the first place and why it was expecting a `CompactionInterruptedException` in the first place (it was failing because that exception was not thrown, which frankly made sense to me). Adding to that, that test didn't feel very useful: it was called "concurrent index build with compaction" but 1) it wasn't even using compaction, just sstable rewrite and 2) it was using barriers in such a way that there wasn't much concurrency happening. Tl;dr, I didn't see how to salvage this into an actually useful test and so my commit just remove it. But it could be worth a look by someone else.

Reply:
@pcmanus Thanks for your analysis.

> I modified the patch to simply document that "disabling index reads" simply means queries will always return an empty result, which should be find in practice.

It works, but in the unlikely case that the property is enabled outside of compactors, or this code path somehow triggers outside of compactors, we would get a silent error. As an alternative, did you consider setting the index as non-queryable (which would throw cause an exception to be thrown)?

> fixing TestWithConcurrentVerification meant that one of the existing test, testConcurrentIndexBuildWithCompaction, started failing. I looked a bit, and tbh I couldn't quite figure what the test was trying to do in the first place and why it was expecting a CompactionInterruptedException in the first place (it was failing because that exception was not thrown, which frankly made sense to me)

The test rightfully expects index creation to interrupt compactions, and this can be seen in the test logs:

```
DEBUG [SecondaryIndexManagement:1] 2024-08-20 11:54:55,897 StorageAttachedIndex.java:424 - [cql_test_keyspace.table_00.table_00_v1_idx] Stopping active compactions to make sure all sstables are indexed after initial build.
```

That said, as you noted, this doesn't cause the "sstable rewrite" compaction to be actually interrupted, I suspect simply because the injection point of the barrier is wrong. Dropping the test for the sake of unblocking the ticket is fine, but maybe we should open a followup to rewrite it properly?

Reply:
> did you consider setting the index as non-queryable

Didn't though of it, but that's a good idea. Pushed a commit to do that and adapt the test accordingly.

>  Dropping the test for the sake of unblocking the ticket is fine, but maybe we should open a followup to rewrite it properly?

I spend 10 more minute looking at it, but I'm just not familiar with compaction enough to figure out how it was supposed to be adapted and I'm giving up (it's 100% unrelated to the changes here and feels hardly urgent). But I've added back the test with `@Ignore` and some quick comment if we think we want to eventually fix it. And happy to create a ticket to follow up, but where do we put this kind of ticket? It's a pure C* thing, do we still use an issue on https://github.com/riptano/cndb? or some JIRA somewhere?